access_log/access_log_record.go

```diff
@@ -39,49 +41,56 @@ func (r *AccessLogRecord) ResponseTime() float64 {
 	return float64(r.FinishedAt.UnixNano()-r.StartedAt.UnixNano()) / float64(time.Second)
 }
 
-func (r *AccessLogRecord) makeRecord() *bytes.Buffer {
-	b := &bytes.Buffer{}
-	fmt.Fprintf(b, `%s - `, r.Request.Host)
-	fmt.Fprintf(b, `[%s] `, r.FormatStartedAt())
-	fmt.Fprintf(b, `"%s %s %s" `, r.Request.Method, r.Request.URL.RequestURI(), r.Request.Proto)
+// memoize makeRecord()
+func (r *AccessLogRecord) getRecord() string {
+	if len(r.record) == 0 {
+		r.record = r.makeRecord()
+	}
+
+	return r.record
+}
 
-	if r.StatusCode == 0 {
-		fmt.Fprintf(b, "MissingResponseStatusCode ")
-	} else {
-		fmt.Fprintf(b, `%d `, r.StatusCode)
+func (r *AccessLogRecord) makeRecord() string {
+	statusCode, responseTime, appId, extraHeaders := "-", "-", "-", ""
+
+	if r.StatusCode != 0 {
+		statusCode = strconv.Itoa(r.StatusCode)
 	}
 
-	fmt.Fprintf(b, `%d `, r.RequestBytesReceived)
-	fmt.Fprintf(b, `%d `, r.BodyBytesSent)
-	fmt.Fprintf(b, `"%s" `, r.FormatRequestHeader("Referer"))
-	fmt.Fprintf(b, `"%s" `, r.FormatRequestHeader("User-Agent"))
-	fmt.Fprintf(b, `%s `, r.Request.RemoteAddr)
-	fmt.Fprintf(b, `x_forwarded_for:"%s" `, r.FormatRequestHeader("X-Forwarded-For"))
-	fmt.Fprintf(b, `x_forwarded_proto:"%s" `, r.FormatRequestHeader("X-Forwarded-Proto"))
-	fmt.Fprintf(b, `vcap_request_id:%s `, r.FormatRequestHeader("X-Vcap-Request-Id"))
-
-	if r.ResponseTime() < 0 {
-		fmt.Fprintf(b, "response_time:MissingFinishedAt ")
-	} else {
-		fmt.Fprintf(b, `response_time:%.9f `, r.ResponseTime())
+	if r.ResponseTime() >= 0 {
+		responseTime = strconv.FormatFloat(r.ResponseTime(), 'f', -1, 32)
```

Comment:
Response time includes precisions up to nanoseconds which can be up to 9 digits after the period.
We would like to avoid losing precision in the access log for the response time field.

A sample case when the above code would lose precision in the response time field: 
http://play.golang.org/p/6hiqyDR37h